src/future/cache.rs

```diff
@@ -1421,170 +1502,238 @@ where
+    ) -> Entry<K, V>
+    where
+        K: Borrow<Q>,
+        Q: ToOwned<Owned = K> + Hash + Eq + ?Sized,
+    {
+        match self.base.get_with_hash(key, hash, true) {
+            Some(entry) => entry,
+            None => {
+                let key = Arc::new(key.to_owned());
+                let value = init();
+                self.insert_with_hash(Arc::clone(&key), hash, value.clone())
+                    .await;
+                Entry::new(Some(key), value, true)
+            }
+        }
+    }
+
+    pub(crate) async fn get_or_optionally_insert_with_hash_and_fun<F>(
         &self,
         key: Arc<K>,
         hash: u64,
         init: F,
-    ) -> Result<V, Arc<E>>
+        need_key: bool,
+    ) -> Option<Entry<K, V>>
     where
-        F: Future<Output = Result<V, E>>,
-        E: Send + Sync + 'static,
+        F: Future<Output = Option<V>>,
     {
-        if let Some(v) = self.base.get_with_hash(&key, hash) {
-            return Ok(v);
+        let entry = self.base.get_with_hash(&key, hash, need_key);
+        if entry.is_some() {
+            return entry;
         }
 
-        self.try_insert_with_hash_and_fun(key, hash, init).await
+        self.optionally_insert_with_hash_and_fun(key, hash, init, need_key)
+            .await
     }
 
-    async fn get_or_try_insert_with_hash_by_ref_and_fun<F, E, Q>(
+    pub(crate) async fn get_or_optionally_insert_with_hash_by_ref_and_fun<F, Q>(
         &self,
         key: &Q,
         hash: u64,
         init: F,
-    ) -> Result<V, Arc<E>>
+        need_key: bool,
+    ) -> Option<Entry<K, V>>
     where
-        F: Future<Output = Result<V, E>>,
-        E: Send + Sync + 'static,
+        F: Future<Output = Option<V>>,
         K: Borrow<Q>,
         Q: ToOwned<Owned = K> + Hash + Eq + ?Sized,
     {
-        if let Some(v) = self.base.get_with_hash(key, hash) {
-            return Ok(v);
+        let entry = self.base.get_with_hash(key, hash, need_key);
+        if entry.is_some() {
+            return entry;
         }
+
         let key = Arc::new(key.to_owned());
-        self.try_insert_with_hash_and_fun(key, hash, init).await
+        self.optionally_insert_with_hash_and_fun(key, hash, init, need_key)
+            .await
     }
 
-    async fn try_insert_with_hash_and_fun<F, E>(
+    async fn optionally_insert_with_hash_and_fun<F>(
         &self,
         key: Arc<K>,
         hash: u64,
         init: F,
-    ) -> Result<V, Arc<E>>
+        need_key: bool,
+    ) -> Option<Entry<K, V>>
     where
-        F: Future<Output = Result<V, E>>,
-        E: Send + Sync + 'static,
+        F: Future<Output = Option<V>>,
     {
         use futures_util::FutureExt;
 
         let get = || {
             let ignore_if = None as Option<&mut fn(&V) -> bool>;
             self.base
-                .get_with_hash_but_no_recording(&key, hash, ignore_if)
+                .get_with_hash_but_no_recording(&key, hash, ignore_if, false)
+                .map(Entry::into_value)
```

Comment:
Perhaps this method does not need `need_key` param and should return `Option<V>`?